src/EnergyPlus/HVACVariableRefrigerantFlow.cc

```diff
@@ -13726,6 +13726,9 @@ void VRFCondenserEquipment::VRFOU_CompSpd(
 
                 } else {
                     CompSpdActual = this->CompressorSpeed(1) * (Q_evap_req * C_cap_operation) / CompEvaporatingCAPSpd(1);
+                    if (Q_cond_req - CompEvaporatingPWRSpd(CounterCompSpdTemp) < 0.0) { // use compressor power to meet condenser required load
+                        CompSpdActual = this->CompressorSpeed(1) * (Q_cond_req * C_cap_operation) / CompEvaporatingPWRSpd(1);
+                    }
                 }
```

Comment:
@yujiex I think I understand the intent here. This could be simpler.
```
}  else {
    if (Q_evap_req < 0.0) {  // use compressor power to meet condenser required load
          CompSpdActual = this->CompressorSpeed(1) * (Q_cond_req * C_cap_operation) /CompEvaporatingPWRSpd(1);
    } else {
          CompSpdActual = this->CompressorSpeed(1) * (Q_evap_req * C_cap_operation) / CompEvaporatingCAPSpd(1);
    }

Reply:
Indeed, this is more clear.

Reply:
What I don't understand is what looks like a COP term = Qc / PWR. I just don't see how that gets back to a speed 1 fraction. What am I missing?

    CompSpdActual = this->CompressorSpeed(1) * (Q_cond_req * C_cap_operation) /CompEvaporatingPWRSpd(1);

Reply:
@rraustad The way I think about this is when `Q_cond_req - CompEvaporatingPWRSpd(1) <= 0.0`, then what `CompEvaporatingPWRSpd(1)` can supply is too much to meet `Q_cond_req`. The speed should be somewhere between 0 and the 1st speed level. So in this case `Q_cond_req` should be met by `CompEvaporatingPWRSpd(1) * ratio`. This `ratio` is calculated as `Q_cond_req / CompEvaporatingPWRSpd(1)`

Reply:
But CompEvaporatingPWRSpd is power, not capacity. So `Q_cond_req / CompEvaporatingPWRSpd(1)` looks strange to me. Shouldn't both equations have CompEvaporatingCAPSpd in the denominator (line 13729)?

    Array1D<Real64> CompEvaporatingPWRSpd; // Array for the compressor power at certain speed [W]
    Array1D<Real64> CompEvaporatingCAPSpd; // Array for the evaporating capacity at certain speed [W]

Reply:
Do you think it should be like this instead when `Q_evap_req < 0`? @rraustad @mjwitte 

CompSpdActual = this->CompressorSpeed(1) * (Q_evap_req * C_cap_operation) / (CompEvaporatingCAPSpd(1) + CompEvaporatingPWRSpd(1));

Reply:
At least this version of the equation made me realize that Q_evap_req = Q_cond_req - PWR. That's what I was missing.

    } else {
        // Capacity to meet is for condenser
        Q_cond_req = Q_req;

So PLR = Q_cond_req / (CompEvaporatingCAPSpd(CounterCompSpdTemp) + CompEvaporatingPWRSpd(CounterCompSpdTemp) where PLR <= 1.

and PLR = ratio so:

    Q_evap_req = Q_cond_req - CompEvaporatingPWRSpd(1) * RTF.

The last term, RTF = PLR/PLF, makes me think that this isn't as simple as it seems to go back and forth between capacity and power since AirConditioner:VariableRefrigerantFlow:FluidTemperatureControl doesn't have a PLFFPLR term so RTF = PLR for this model.

Reply:
I think with the suggestion above that old line 13728 will be changed to this and you won't need the new IF block.

    CompSpdActual = this->CompressorSpeed(1) * PLR

Reply:
I also think line 13715 would simply be:

    if (PLR <= 1.0) {

Reply:
Should `PLR` be `Q_cond_req / (CompEvaporatingCAPSpd(CounterCompSpdTemp) / C_cap_operation + CompEvaporatingPWRSpd(CounterCompSpdTemp));
`? (as there seems to be this `C_cap_operation` term in other places to adjust for evaporative load.

If I understand correctly, this is how the code should look like? @rraustad 
```
CompEvaporatingPWRSpd(CounterCompSpdTemp) =
    this->RatedCompPower * CurveValue(state, this->OUCoolingPWRFT(CounterCompSpdTemp), T_discharge, T_suction);
CompEvaporatingCAPSpd(CounterCompSpdTemp) =
    this->CoffEvapCap * this->RatedEvapCapacity * CurveValue(state, this->OUCoolingCAPFT(CounterCompSpdTemp), T_discharge, T_suction);
Real64 PLR = Q_cond_req / (CompEvaporatingCAPSpd(CounterCompSpdTemp) / C_cap_operation + CompEvaporatingPWRSpd(CounterCompSpdTemp));
Q_evap_req = Q_cond_req - CompEvaporatingPWRSpd(CounterCompSpdTemp) * PLR;

if (PLR <= 1.0) {
    // Compressor speed stage CounterCompSpdTemp need not to be increased, finish Iteration DoName1

    if (CounterCompSpdTemp > 1) {

        CompSpdLB = CounterCompSpdTemp - 1;
        CompSpdUB = CounterCompSpdTemp;

        CompSpdActual = this->CompressorSpeed(CompSpdLB) + (this->CompressorSpeed(CompSpdUB) - this->CompressorSpeed(CompSpdLB)) /
                                                                (CompEvaporatingCAPSpd(CompSpdUB) - CompEvaporatingCAPSpd(CompSpdLB)) *
                                                                PLR;

    } else {
        CompSpdActual = this->CompressorSpeed(1) * PLR;
    }

    break; // EXIT DoName1
}
```

Reply:
That looks better. You don't need Q_evap_rated here. Just use PLR as the multiplier for speed 1 and speed > 1. 

CompSpdActual = this->CompressorSpeed(CompSpdLB) + (this->CompressorSpeed(CompSpdUB) - this->CompressorSpeed(CompSpdLB)) * PLR;

Now that you mention the C_cap_operation term it does seem likes it's necessary. I am just wondering how to apply it. There is a CapFT term and an EIRFT term that might need to be applied to Cap and Pwr, respectively. Think about that a little more. This is getting closer.

Reply:
Oh right, the PLR > 1 case compressor speed should be `CompSpdActual = this->CompressorSpeed(CompSpdLB) + (this->CompressorSpeed(CompSpdUB) - this->CompressorSpeed(CompSpdLB)) * PLR;`

For the CapFT and EIRFT terms, I don't think the FluidTCtrl models have those curves. They have just the two curves: "Evaporative Capacity Multiplier Function of Temperature Curve Name" (CAPSpd) and "Compressor Power Multiplier Function of Temperature Curve Name" (PWRSpd). They don't have the EIR curve. The CapFT loos like the CAPSpd's role maybe?

Reply:
The reason why I felt the `C_cap_operation` should be applied to the `CompEvaporatingCAPSpd(CounterCompSpdTemp)` is because previously, `Q_evap_req * C_cap_operation` was supposed to match `CompEvaporatingCAPSpd(CounterCompSpdTemp)`, so it seems `Q_evap_req` might need to match with some term like `CompEvaporatingCAPSpd(CounterCompSpdTemp) / C_cap_operation`. Also according to the comment in the function computing `C_cap_operation` (`VRFOU_CapModFactor`), "This factor is used to modify the system evaporative capacity, by describing the difference between rated conditions and real conditions", so evaporative capacity need to have this adjustment maybe?

Reply:
If I look at the equations for compressor speed and power in VRFOU_CalcCompC and they look similar to what was here before. I say that because I don't what to change the original intent too much. From what I see in these equations, Q_evap_coil = Q_evap_req * C_cap_operation. And power looks to be proportional to compressor speed. So even though these equations look different I think they are doing the same thing. I'm not really sure that C_cap_operation is used correctly here since I think Q_evap_req is the coil load and it may be that C_cap_operation should not be applied here. It's easy to get confused between applying the CapFT term to the rated capacity versus the coil load. I think what you have here looks pretty close and testing needs to be performed to confirm. The only thing I'm not sure is missing is the adjustment for power at off-rated conditions at new line 13712. This term: `CurveValue(state, this->OUCoolingPWRFT(CounterCompSpdTemp), T_discharge, T_suction)`

    CompSpdActual = this->CompressorSpeed(CompSpdLB) + (this->CompressorSpeed(CompSpdUB) - this->CompressorSpeed(CompSpdLB)) /
                                                           (CompEvaporatingCAPSpd(CompSpdUB) - CompEvaporatingCAPSpd(CompSpdLB)) *
                                                           (Q_evap_req * C_cap_operation - CompEvaporatingCAPSpd(CompSpdLB));

    Ncomp = CompEvaporatingPWRSpd(CompSpdLB) + (CompEvaporatingPWRSpd(CompSpdUB) - CompEvaporatingPWRSpd(CompSpdLB)) /
                                                   (this->CompressorSpeed(CompSpdUB) - this->CompressorSpeed(CompSpdLB)) *
                                                   (CompSpdActual - this->CompressorSpeed(CompSpdLB));

Reply:
I think the equation at new line 13712 should look like this since these data are rated values (please check me on this):

    Real64 PLR = Q_cond_req / (CompEvaporatingCAPSpd(CounterCompSpdTemp) * C_cap_operation +
                          CompEvaporatingPWRSpd(CounterCompSpdTemp) *
                          CurveValue(state, this->OUCoolingPWRFT(CounterCompSpdTemp), T_discharge, T_suction));

Reply:
You don't need this at new line 13713 since the following equations no longer use Q_evap_req:

     Q_evap_req = Q_cond_req - CompEvaporatingPWRSpd(CounterCompSpdTemp) * PLR;

Reply:
right. I'll remove it.

Reply:
I don't know the original intent but Tsuction + 8 might be the rated superheat and Tdischarge - 5 the rated subcool. But I am just guessing and not sure that even makes sense.

Reply:
That might be about right.

I just did a quick plot of the C_cap_operation distribution in `VRFOU_CompSpd`. Looks like it's a pretty substantial adjustment factor.

![image](https://github.com/user-attachments/assets/5f3cecca-7f16-4d49-8013-85da3d4096ef)

Reply:
I see at line 13708 - 13711 are these equations. These are already multiplied by the modifier curves.

            CompEvaporatingPWRSpd(CounterCompSpdTemp) =
                this->RatedCompPower * CurveValue(state, this->OUCoolingPWRFT(CounterCompSpdTemp), T_discharge, T_suction);
            CompEvaporatingCAPSpd(CounterCompSpdTemp) =
                this->CoffEvapCap * this->RatedEvapCapacity * CurveValue(state, this->OUCoolingCAPFT(CounterCompSpdTemp), T_discharge, T_suction);

I think that means this equation is correct:

    PLR = Q_cond_req / (CompEvaporatingCAPSpd(CounterCompSpdTemp) + 
             CompEvaporatingPWRSpd(CounterCompSpdTemp))

Reply:
I looked for C_cap_operating and see that this is a density ratio * enthalpy ratio calculated in VRFOU_CapModFactor. So yes, I think this is needed to multiply the CompEvaporatingCAPSpd(CounterCompSpdTemp) term above.

    C_cap_operation = C_cap_density * C_cap_enthalpy;

Reply:
From what I can see from the VRF calculations, C_cap_operation is an adjustment of the evaporator load to find the compressor load. These calculations revolve around finding the load on the compressor, using the evaporator load, which is always upstream of the compressor, whether in cooling or heating mode. I'm worried that I might be leading you in the wrong direction with this model. I think C_cap_operatoin is only applied to the evaporator load. Does that mean that Q_evap_req DOES need to be calculated here? I'm not sure.

Reply:
Also be aware that the developer of this model had the help of Daikin manufacturer engineers to create these equations.

Reply:
> I looked for C_cap_operating and see that this is a density ratio * enthalpy ratio calculated in VRFOU_CapModFactor. So yes, I think this is needed to multiply the CompEvaporatingCAPSpd(CounterCompSpdTemp) term above.
> 
> ```
> C_cap_operation = C_cap_density * C_cap_enthalpy;
> ```

Should the `C_cap_enthalpy` be applied by `CompEvaporatingCAPSpd(CounterCompSpdTemp) /  C_cap_operation`? As the `C_cap_operation` is rated divided by real, so to get real, you need to divide by it (I assume the `CompEvaporatingCAPSpd(CounterCompSpdTemp)` is about rated) ?

Reply:
There might be some issue of applying the PLR to the CounterCompSpdTemp > 1 case, as in the original code, the scaling factor correspond to the part between the LB and UB. So to use in the CounterCompSpdTemp > 1 case, they probably want something like
```
PLR = Q_cond_req / (CompEvaporatingCAPSpd(CounterCompSpdTemp) - CompEvaporatingCAPSpd(CounterCompSpdTemp - 1) + CompEvaporatingPWRSpd(CounterCompSpdTemp) - CompEvaporatingPWRSpd(CounterCompSpdTemp - 1));
```

CounterCompSpdTemp > 1 and CounterCompSpdTemp == 0 will have different forms of PLR in this case.

Reply:
I did some derivation based on the current code

```
in the original code
Q_evap_req = Q_cond_req - CompEvaporatingPWRSpd(CounterCompSpdTemp);
I think the intent was
Q_cond_req = Q_req = Q_evap_req + compressor heat
but compressor heat might not actually be CompEvaporatingPWRSpd(CounterCompSpdTemp)

let
k be the speed level
C be the short form for C_cap_operation
PWR(.) be the short form of CompEvaporatingPWRSpd(.)
CAP(.) be the short form of CompEvaporatingCAPSpd(.)
spd(.) be the short form of this->CompressorSpeed(.)
deltaPWR be PWR(k) - PWR(k - 1)
deltaCAP be CAP(k) - CAP(k - 1)

I think compressor heat should be this instead of just PWR(k)
compressor heat = deltaPWR * r + PWR(k - 1)
so
Q_cond_req = Q_req = Q_evap_req + deltaPWR * r + PWR(k - 1) <---- eq 1

we also know this from the CompSpdActual calculation equation that
CompSpdActual = Spd(k - 1) + deltaSpd / deltaCAP * (Q_evap_req * C - CAP(k - 1))
so we call the following r
r = (Q_evap_req * C - CAP(k - 1)) / deltaCAP
so
CompSpdActual = Spd(k - 1) + deltaSpd * r

arranging terms in eq 1
Q_cond_req - deltaPWR * r + PWR(k - 1) = Q_evap_req
(Q_cond_req - deltaPWR * r + PWR(k - 1)) * C - CAP(k - 1) = Q_evap_req * C - CAP(k - 1)
((Q_cond_req - deltaPWR * r + PWR(k - 1)) * C - CAP(k - 1)) / deltaCAP = (Q_evap_req * C - CAP(k - 1)) / deltaCAP
((Q_cond_req - deltaPWR * r + PWR(k - 1)) * C - CAP(k - 1)) / deltaCAP = r
(Q_cond_req - deltaPWR * r + PWR(k - 1)) * C - CAP(k - 1) = deltaCAP * r
(Q_cond_req + PWR(k - 1)) * C - CAP(k - 1) = deltaCAP * r + deltaPWR * r * C
so
r = ((Q_cond_req + PWR(k - 1)) * C - CAP(k - 1)) / (deltaCAP + deltaPWR * C)
  = ((Q_cond_req + PWR(k - 1)) - CAP(k - 1)/C) / (deltaCAP/C + deltaPWR)

in the special case where k = 1, then k - 1, PWR(k - 1) and CAP(k - 1) will all be 0

r = ((Q_cond_req) * C) / (CAP(1) + PWR(1) * C)
  = (Q_cond_req) / (CAP(1)/C + PWR(1))
```

Reply:
This looks reasonable and starts out good:

    I think compressor heat should be this instead of just PWR(k)
    compressor heat = deltaPWR * r + PWR(k - 1)
    so
    Q_cond_req = Q_req = Q_evap_req + deltaPWR * r + PWR(k - 1) <---- eq 1

But there should be a minus sign on the lower speed power when Eq 1 got rearranged.

    Q_cond_req - deltaPWR * r   "-"   PWR(k - 1) = Q_evap_req  <---- eq 2

I think it boils down to correcting the sign in the last equation as:

    r = ((Q_cond_req  "-"  PWR(k - 1)) * C - CAP(k - 1)) / (deltaCAP + deltaPWR * C)
      = ((Q_cond_req  "-"  PWR(k - 1)) - CAP(k - 1)/C) / (deltaCAP/C + deltaPWR)

So that would be a rather big equation when expanding Eq 2 just to get to Q_evap_req as used in the original code. That's OK but it seems like a lot of work to get there. And CAP/C seems strange based on the original model so what does CAP/C physically represent? C in this model seems to represent a different load based on system operation at off design conditions, so then what does CAP/C mean? All I have is questions. If you want to try this equation to see how the model responds then maybe that would help determine if this has any merit. Also, I would think the compressor load would be higher than the coil load and in [your figure](https://github.com/NREL/EnergyPlus/pull/10649#discussion_r1809698105) you showed C_cap_operation < 1 at 0.6, I'm still trying to figure that one out. But if C is < 1 then CAP/C would show a higher compressor load than at the coil, which seems to make physical sense. Or is this input file just operating at off design conditions? What if the input file operated at the other side of the design conditions, would C be > 1?

I do think your right in that the original equation is wrong because it doesn't account for the fraction of the upper compressor speed.

    Q_evap_req = Q_cond_req - CompEvaporatingPWRSpd(CounterCompSpdTemp);

If you could somehow use the Spd terms, both CAP and PWR to determine what speed to operate at you would solve part of this problem. I still like the PLR attempt just not sure it's accurate. For example:

    if (Q_cond_req < CompEvaporatingCAPSpd(.) + CompEvaporatingPWRSpd(.)) {
        you found the speed, now figure out the compressor speed fraction,  but I think you might still be stuck with your derivation here
    }

Reply:
Good catch! There should be a negative sign for the `PWR(k - 1)`.

For the CAP/C, I think CAP is about the capacity at rated condition, and C is some multiplier equals roughly rated-condition/real-condition, so CAP/C will be the capacity at real condition?

I'll check out if the C_cap_operation is larger than 1 on the other side of the rating condition or it never reaches over 1 because of the equation setup (I felt in my previous debugging prints of other stuff this term might have not shown larger than 1, but I'll double check).

The PLR way is definitely more elegant, I'll see if I can reduce the calculation and simplify it.

Reply:
Coil capacity at off-rated conditions is:

            CompEvaporatingCAPSpd(CounterCompSpdTemp) =
                this->CoffEvapCap * this->RatedEvapCapacity * CurveValue(state, this->OUCoolingCAPFT(CounterCompSpdTemp), T_discharge, T_suction);

C_cap_operation is a different modifier and has to do with the refrigerant side in VRFOU_CapModFactor.

Reply:
I see. Then it does seem strange that CAP is dividing C

Reply:
Yes. In this equation it looks like it's calculating the operating coil capacity divided by the off-design adjusted coil capacity at speed 1. That's what I was trying to do with PLR, I just don't understand C_cap_operation yet. If Q_evap_req is the coil load, the load the TU want's to meet, then what is this C_cap_operation trying to do ??? i.e., coil load / coil capacity = PLR. But this equation is calculating a compressor parameter, not a coil parameter. If C_cap_operation was closer to 1 I think it would be adjusting for piping losses, but it's 0.6. If piping losses were 40% of the coil load they wouldn't sell any products. So what is it?

     CompSpdActual = this->CompressorSpeed(1) * (Q_evap_req * C_cap_operation) / CompEvaporatingCAPSpd(1);

Reply:
![image](https://github.com/user-attachments/assets/a5316abc-b7b2-4749-9744-7493d06d9378)
I just plotted the compressor speed vs coil load (Q_evap_req). The left subplot is in the develop branch, still with the negative compressor speed issue. The right is this current branch. I used the newly derived ratio r in the calculation

Reply:
That does look promising, doesn't it. Is CompSpdActual 0 - 400 over multiple speeds? Or is this for speed 1? And what is the maximum comp speed at speed 1? I'm trying to get a feel for if this is across the entire range of the compressor speeds.

Reply:
just updated the plot a bit, with longer duration

Reply:
the compressor speed for the lowest speed level (loading index 1) is 1500.

Reply:
I can also try to change the VRF evaporative capacity to make it reach higher speed levels.

Reply:
This is how they look when they can reach higher speed levels (I reduced the capacity to 10000, the autosized heating and cooing capacity are around 70000W)

![image](https://github.com/user-attachments/assets/415008b2-6c6b-4c5e-bf5b-836302051898)

Reply:
Can you make it span speeds? like from speed 1 to speed 3? I want to see what Q_evap_req * C_cap_operating / CompEvaporatingCAPSpd(x) looks like (or your derivation). I suspect there are discontinuities because C_cap_operatiion = 0.6. And make a scatter plot like before. Q_evap_req versus compressor speed.

Reply:
Oh I see, your plot is Q_evap_req versus comp speed. I was confused because I didn't see a 45 degree line like I was expecting. Why is Q_evap_req = 2000 at comp speeds 0 - 10000 ??

Reply:
Somehow there's some weird thing happening when I plot with excel. CompSpdActual is always below 6000 for both the feature and develop branch output, but it somehow plotted it out to over 10000. I have attached the data here
[plot data.zip](https://github.com/user-attachments/files/17510759/plot.data.zip)
I used R to plot, this is how they look

![image](https://github.com/user-attachments/assets/77e0c04b-631b-48c4-b981-69d4045952e9)